src/sfizz/FileMetadata.cpp

```diff
@@ -0,0 +1,396 @@
+
+    if (count >= 4 && !memcmp(magic, "fLaC", 4)) {
+        if (!impl_->openFlac()) {
+            close();
+            return false;
+        }
+    }
+    else if (count >= 4 && !memcmp(magic, "RIFF", 4)) {
+        if (!impl_->openRiff()) {
+            close();
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void FileMetadataReader::close()
+{
+    impl_->stream_.reset();
+    impl_->riffChunks_.clear();
+}
+
+bool FileMetadataReader::Impl::openFlac()
+{
+    FILE* stream = stream_.get();
+    std::vector<RiffChunkInfo>& riffChunks = riffChunks_;
+
+    if (fseek(stream, 4, SEEK_SET) != 0)
+        return false;
+
+    uint32_t header = 0;
+    while (((header >> 31) & 1) != 1) {
+        if (!fread_u32be(stream, header))
+            return false;
+
+        const uint32_t blockType = (header >> 24) & 0x7f;
+        const uint32_t blockSize = header & ((1 << 24) - 1);
+
+        const off_t offStartBlock = ftell(stream);
+        const off_t offNextBlock = offStartBlock + blockSize;
+
+        if (blockType == 2) { // APPLICATION block
+            char blockId[4];
+            char riffId[4];
+            uint32_t riffChunkSize;
+            if (fread(blockId, 4, 1, stream) == 1 && memcmp(blockId, "riff", 4) == 0 &&
+                fread(riffId, 4, 1, stream) == 1 &&
+                fread_u32le(stream, riffChunkSize) && riffChunkSize <= blockSize - 12)
+            {
+                RiffChunkInfo info;
+                info.index = riffChunks.size();
+                info.fileOffset = ftell(stream);
+                memcpy(info.id.data(), riffId, 4);
+                info.length = riffChunkSize;
+                riffChunks.push_back(info);
+            }
+        }
+
+        if (fseek(stream, offNextBlock, SEEK_SET) != 0)
+            return false;
+    }
+
+    return true;
+}
+
+bool FileMetadataReader::Impl::openRiff()
+{
+    FILE* stream = stream_.get();
+    std::vector<RiffChunkInfo>& riffChunks = riffChunks_;
+
+    if (fseek(stream, 12, SEEK_SET) != 0)
+        return false;
+
+    char riffId[4];
+    uint32_t riffChunkSize;
+    while (fread(riffId, 4, 1, stream) == 1 && fread_u32le(stream, riffChunkSize)) {
+        RiffChunkInfo info;
+        info.index = riffChunks.size();
+        info.fileOffset = ftell(stream);
+        memcpy(info.id.data(), riffId, 4);
+        info.length = riffChunkSize;
+        riffChunks.push_back(info);
+
+        if (fseek(stream, riffChunkSize, SEEK_CUR) != 0)
+            return false;
+    }
+
+    return true;
+}
+
+size_t FileMetadataReader::riffChunkCount() const
+{
+    return impl_->riffChunks_.size();
+}
+
+const RiffChunkInfo* FileMetadataReader::riffChunk(size_t index) const
+{
+    return impl_->riffChunk(index);
+}
+
+const RiffChunkInfo* FileMetadataReader::Impl::riffChunk(size_t index) const
+{
+    const std::vector<RiffChunkInfo>& riffChunks = riffChunks_;
+    return (index < riffChunks.size()) ? &riffChunks[index] : nullptr;
```

Comment:
I didn't try, but a static cast to `const RiffChunkInfo*` is not working?

Reply:
Sorry explain what is the problem please, I don't understand it.

Reply:
Not really a problem per-say. I'm thinking you're creating the const ref to get a const pointer from the original vector, and I'm wondering if for some reason it's not possible to just cast to `const RiffChunkInfo*`. But it's not blocking, just curiosity. I will just try on my copy :D

Reply:
```c++
const RiffChunkInfo* FileMetadataReader::Impl::riffChunk(size_t index) const
{
    return (index < riffChunks_.size()) ? &riffChunks_[index] : nullptr;
}
```
This seems to compile for me.

Reply:
Oh you prefer the one-line, sure I'll change it.